src/tfs/tlog.c

```diff
@@ -202,31 +440,89 @@ closure_function(2, 1, void, log_read_complete,
     tl->dictionary = newdict;
 #endif
 
-    apply(sh, 0);
+  out_apply_status:
+    tlog_debug("log_read_complete exit with status %v\n", s);
+    buffer_clear(tl->tuple_staging);
+    apply(sh, s);
+  out:
     closure_finish();
 }
 
-void read_log(log tl, u64 offset, u64 size, status_handler sh)
+static boolean init_staging(log tl, status_handler sh)
 {
+    char *err;
+    u64 size = log_size(tl);
     tl->staging = allocate_buffer(tl->h, size);
+    if (tl->staging == INVALID_ADDRESS) {
+        err = "failed to allocate staging buffer";
+        goto fail;
+    }
+    tlog_debug("reading log extension, sectors %R, staging %p\n",
+               tl->sectors, tl->staging);
+
+#ifndef BOOT
```

Comment:
maybe hanging off a more general writable flag? idk

Reply:
Yeah, these could just be gated by the presence of tl->fs->w. Though in this case stage2 doesn't link in the id heap...this should probably be hidden in tfs.c.